
fef68415c8adf4d24557a82cb38aea389d72b213

diff --git a/examples/connection.rs b/examples/connection.rs
index f961046..90677fd 100644
--- a/examples/connection.rs
+++ b/examples/connection.rs
@@ -1,15 +1,18 @@
 extern crate my_mio;
 extern crate bytes;
+extern crate parking_lot;
 
 use std::{io, mem, fmt};
 use my_mio::{Events, Poll, PollOpt, Ready, Token, Job};
 use bytes::{Buf, BufMut, Bytes, BytesMut};
 use my_mio::net::{TcpListener, TcpStream, EventLoop, EventLoopBuilder, Acceptor, TcpConnection};
 use std::net::{self, SocketAddr, SocketAddrV4, SocketAddrV6, Ipv4Addr, Ipv6Addr};
-use std::sync::{Arc, Mutex, Condvar};
+use std::sync::{Arc, Condvar};
+use parking_lot::{Mutex, ReentrantMutex};
 use std::sync::atomic::{AtomicUsize, AtomicPtr, AtomicBool};
 use std::sync::atomic::Ordering::{self, Acquire, Release, AcqRel, Relaxed, SeqCst};
 use std::time::Duration;
+use std::ops::Deref;
 
 fn sleep_ms(ms: u64) {
     use std::thread;
@@ -22,11 +25,10 @@ const CLIENT: Token = Token(10_000_000);
 fn default_accept_cb(stream: TcpStream, addr: SocketAddr) {
     println!("accept {}", addr);
     let event_loop = EventLoopBuilder::get_current_loop();
-    let mut conn = Arc::new(Mutex::new(TcpConnection::new(event_loop.clone(), stream)));
+    let mut conn = Arc::new(std::sync::Mutex::new(TcpConnection::new(event_loop.clone(), stream)));
     let clone = conn.clone();
     let job = Box::new(move |val: i64| { clone.lock().unwrap().handleEvent(val); });
-    println!("event_loop.lock: {}", event_loop.is_poisoned());
-    event_loop.lock().unwrap().register(&conn.lock().unwrap().sock, CLIENT, Ready::readable(), 
+    event_loop.lock().deref().borrow().register(&conn.lock().unwrap().sock, CLIENT, Ready::readable(), 
             PollOpt::edge(), job);
 }
 
@@ -39,7 +41,7 @@ fn main() {
         .timer_capacity(65536);
     let mut event_loop = b.get_build().unwrap();
     let mut acceptor = Arc::new (
-        Mutex::new (
+        std::sync::Mutex::new (
             Acceptor::new(event_loop.clone(), &"0.0.0.0:9527".to_string())
         )
     );
@@ -48,10 +50,6 @@ fn main() {
     let job = Box::new(move |val: i64| { clone.lock().unwrap().handleRead(val); });
     acceptor.lock().unwrap().bind(job);
     acceptor.lock().unwrap().set_accept_cb(default_accept_cb);
-
-    let ptr: *mut Mutex<EventLoop> = Arc::as_ptr(&mut event_loop) as *mut _;
-    unsafe {
-        (*ptr).get_mut().unwrap().run();
-    }
+    event_loop.lock().deref().borrow_mut().run();
 }
 
diff --git a/examples/remutex.rs b/examples/remutex.rs
index 3f6df16..1c20456 100644
--- a/examples/remutex.rs
+++ b/examples/remutex.rs
@@ -39,6 +39,10 @@ impl User {
     fn run(&mut self) {
         let user = get_current_user();
         //println!("is_poisoned: {}", user.is_poisoned());
+        //let id = user.lock().deref().borrow_mut().get_id(); // already borrowed: BorrowMutError
+        let id = user.lock().deref().borrow().get_id();
+    
+        println!("run id: {}", id);
         sleep_ms(1000 * 3);
     }
 }
@@ -65,7 +69,7 @@ fn test0() {
     //CURRENT_USER.set(user.clone());
 
     let locked = user.lock(); 
-    let mut u = locked.deref(); // fn deref(&self) -> &T // 返回共享引用
+    let u = locked.deref(); // fn deref(&self) -> &T // 返回共享引用
     let id = u.get_id();
     println!("id: {}", id);
 
diff --git a/src/lib.rs b/src/lib.rs
index 6e62a53..566c5a0 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -4,6 +4,7 @@ extern crate slab;
 extern crate net2;
 extern crate iovec;
 extern crate bytes;
+extern crate parking_lot;
 
 mod event_imp;
 pub use event_imp:: {
diff --git a/src/net/acceptor.rs b/src/net/acceptor.rs
index c380194..fe69dac 100644
--- a/src/net/acceptor.rs
+++ b/src/net/acceptor.rs
@@ -1,6 +1,9 @@
 use net::{EventLoop, TcpStream, TcpListener};
 use std::net::{self, SocketAddr, SocketAddrV4, SocketAddrV6, Ipv4Addr, Ipv6Addr};
-use std::sync::{Arc, Mutex, Condvar};
+use std::sync::{Arc, Condvar};
+use parking_lot::{Mutex, ReentrantMutex};
+use std::cell::{RefCell};
+use std::ops::Deref;
 use {Events, Poll, PollOpt, Ready, Token, Job};
 
 const SERVER: Token = Token(10_000_000);
@@ -10,7 +13,7 @@ unsafe impl Sync for Acceptor {}
 
 pub struct Acceptor {
     tcp_listener: TcpListener,
-    event_loop: Arc<Mutex<EventLoop>>,
+    event_loop: Arc<ReentrantMutex<RefCell<EventLoop>>>,
     is_listening: bool,
     accept_cb: fn(TcpStream, SocketAddr)
 }
@@ -18,7 +21,7 @@ pub struct Acceptor {
 fn default_accept_cb(stream: TcpStream, addr: SocketAddr) {}
 
 impl Acceptor {
-    pub fn new(event_loop: Arc<Mutex<EventLoop>>, addr: &String) -> Acceptor {
+    pub fn new(event_loop: Arc<ReentrantMutex<RefCell<EventLoop>>>, addr: &String) -> Acceptor {
         Acceptor {
             tcp_listener: TcpListener::new(&(addr.parse().unwrap())),
             event_loop: event_loop,
@@ -36,7 +39,7 @@ impl Acceptor {
         (self.accept_cb)(stream, addr);
     }
     pub fn bind(&mut self, job: Job) {
-        self.event_loop.lock().unwrap().register(&self.tcp_listener, SERVER, Ready::readable(), 
+        self.event_loop.lock().deref().borrow().register(&self.tcp_listener, SERVER, Ready::readable(), 
                 PollOpt::edge(), job);
         self.is_listening = true;
     }
diff --git a/src/net/connection.rs b/src/net/connection.rs
index 5e05de3..25783e0 100644
--- a/src/net/connection.rs
+++ b/src/net/connection.rs
@@ -5,10 +5,13 @@ use {Events, Poll, PollOpt, Ready, Token, Job};
 use event_imp::{ready_from_usize, ready_as_usize};
 use net::{EventLoop, TcpStream};
 use std::net::{self, SocketAddr, SocketAddrV4, SocketAddrV6, Ipv4Addr, Ipv6Addr};
-use std::sync::{Arc, Mutex, Condvar};
+use std::sync::{Arc, Condvar};
 use std::time::Duration;
 use std::sync::atomic::{AtomicUsize, AtomicPtr, AtomicBool};
 use std::sync::atomic::Ordering::{self, Acquire, Release, AcqRel, Relaxed, SeqCst};
+use parking_lot::{Mutex, ReentrantMutex};
+use std::cell::{RefCell};
+use std::ops::Deref;
 
 // token incr TODO
 
@@ -28,7 +31,7 @@ pub struct TcpConnection {
     token: Option<Token>,
     interest: Ready,
 
-    event_loop: Arc<Mutex<EventLoop>>,
+    event_loop: Arc<ReentrantMutex<RefCell<EventLoop>>>,
     pub sock: TcpStream,
     // timer
     read_buf: Option<BytesMut>,
@@ -51,7 +54,7 @@ fn default_written_cb() -> bool { false }
 fn default_err_cb() {}
 
 impl TcpConnection {
-    pub fn new(event_loop: Arc<Mutex<EventLoop>>, sock: TcpStream) -> TcpConnection {
+    pub fn new(event_loop: Arc<ReentrantMutex<RefCell<EventLoop>>>, sock: TcpStream) -> TcpConnection {
         TcpConnection {
             token: None,
             interest: Ready::empty(),
@@ -84,7 +87,7 @@ impl TcpConnection {
             Ok(Some(r)) => {
                 println!("Conn: read {} bytes, {:?}", r, buf);
                 if r == 0 {
-                    self.event_loop.lock().unwrap().deregister(&self.sock);
+                    self.event_loop.lock().deref().borrow().deregister(&self.sock);
                 }
                 // buf toto
                 (self.read_cb)(&mut buf);
diff --git a/src/net/event_loop.rs b/src/net/event_loop.rs
index 157238a..ad0311c 100644
--- a/src/net/event_loop.rs
+++ b/src/net/event_loop.rs
@@ -6,8 +6,10 @@ use std::{io, usize};
 use std::default::Default;
 use std::time::Duration;
 use std::{fmt, error, any};
-use std::sync::{Arc, Mutex, Condvar};
+use std::sync::{Arc, Condvar};
+use parking_lot::{Mutex, ReentrantMutex};
 use std::thread_local;
+use std::cell::{RefCell};
 
 pub enum NotifyError<T> {
     Io(io::Error),
@@ -245,10 +247,8 @@ pub struct EventLoopBuilder {
     config: Config,
 }
 
-use std::cell::Cell;
-use std::cell::RefCell;
 thread_local! {
-    pub static current_loop: RefCell<Arc<Mutex<EventLoop>>> = panic!("!"); //Arc::new(Mutex::new(EventLoop));
+    pub static current_loop: RefCell<Arc<ReentrantMutex<RefCell<EventLoop>>>> = panic!("!");
 }
 
 impl EventLoopBuilder {
@@ -278,21 +278,21 @@ impl EventLoopBuilder {
 	pub fn build(self) -> io::Result<EventLoop> {
         EventLoop::configured(self.config)
     }
-    pub fn get_build(self) -> io::Result<Arc<Mutex<EventLoop>>> {
+    pub fn get_build(self) -> io::Result<Arc<ReentrantMutex<RefCell<EventLoop>>>> {
         /*
         if current_loop.try_with() == panic!("!") {
             println!("slkdfjlskdfjl");
         }
         */
-        let event_loop = Arc::new(Mutex::new(self.build().unwrap()));
+        let event_loop = Arc::new(ReentrantMutex::new(RefCell::new(self.build().unwrap())));
         let clone = event_loop.clone();
         current_loop.set(clone);
         Ok(event_loop)
     }
-    pub fn get_current_loop() -> Arc<Mutex<EventLoop>> {
+    pub fn get_current_loop() -> Arc<ReentrantMutex<RefCell<EventLoop>>> {
         //current_loop.with(|poll| -> Arc<Mutex<EventLoop>> {return poll.into_inner()})
         //current_loop.with(|poll| -> &'static mut Arc<Mutex<EventLoop>> {return poll.get_mut()})
-        let ptr = current_loop.with(|poll| -> *mut Arc<Mutex<EventLoop>> {return poll.as_ptr()});
+        let ptr = current_loop.with(|poll| -> *mut Arc<ReentrantMutex<RefCell<EventLoop>>> {return poll.as_ptr()});
         unsafe {
             let clone = (*ptr).clone();
             clone
@@ -305,8 +305,3 @@ impl EventLoopBuilder {
     }
 }
 
-#[derive(Clone)]
-struct EventLoopImp {
-    inner: Arc<Mutex<EventLoop>>,
-}
-
